### DISCUSSION TITLE
General development

### DISCUSSION THREAD

**Discussion by aregtech** [aregtech/areg-sdk#56] [General]
General development discussions.

**Comment by Ali-Nasrolahi** [comment]
Hello @aregtech, thank you for your invitation; it was an awesome experience working with you and @biwiki on this project.

As for merging, I'd be honored to do so.
I created #57 PR .
(Although I'm not a merge guy 😄 , so I don't know if you meant this or not ).
Thanks again.

**Reply by aregtech** [reply]
Hi @Ali-Nasrolahi, same for me :)

Actually, there is nothing special. 
You already merge `20220205-issue18-cmake` into the `20220701-candidate` :)

- your working branch is `20220205-issue18-cmake` (can create another if you want)
- Each time you make changes in `20220205-issue18-cmake` and want to bring them to `20220701-candidate`
**Step 1:** take changes from `20220701-candidate`, resolve conflicts (if any) and push into `20220205-issue18-cmake`  (command to merge to your active branch ` $ git merge 20220701-candidate` )
**Step 2:** create pull-request to bring changes from `20220205-issue18-cmake` into `20220701-candidate`.

Important is to update your `20220205-issue18-cmake` and build projects before bringing changes to `20220701-candidate`.

**Reply by Ali-Nasrolahi** [reply]
Oh, right, now I get it :)

**Reply by aregtech** [reply]
If you have difficulties to update `20220205-issue18-cmake`, let me know. I'll do it.

**Reply by Ali-Nasrolahi** [reply]
Appreciate it. :)

**Comment by aregtech** [comment]
[About merging](https://www.atlassian.com/git/tutorials/merging-vs-rebasing)

> The first thing to understand about git rebase is that it solves the same problem as git merge. Both of these commands are designed to integrate changes from one branch into another branch—they just do it in very different ways.
> 
> Consider what happens when you start working on a new feature in a dedicated branch, then another team member updates the main branch with new commits. This results in a forked history, which should be familiar to anyone who has used Git as a collaboration tool.
> 
> Now, let’s say that the new commits in main are relevant to the feature that you’re working on. To incorporate the new commits into your feature branch, you have two options: merging or rebasing.
>

**Comment by aregtech** [comment]
Hi @Ali-Nasrolahi,
I tried to compile with CMake in [Visual Studio Code](https://code.visualstudio.com/) on Windows machine. There were a few failures. Would it be possible to setup a video call session and talk a little bit? I have couple of open points to clarify.

**Reply by Ali-Nasrolahi** [reply]
Hey @aregtech.
Sure.
When do you prefer to arrange this?

And could you provide me output of `cmake --build` please? So I can debug it, in the meantime.

Thanks

**Reply by aregtech** [reply]
There were several errors related with compiler options. I tried to fix, but then thought better to consult with you.
About the session, can we do it about 14:00 GMT+1:00 or late evening?

**Reply by Ali-Nasrolahi** [reply]
Could you please remove / comment these lines which located in `conf/common.cmake` and try `cmake --build` again.

```cmake
# Add compiler flags here
#add_compile_options(-g -pthread -Werror) line 20

#add_compile_options(-O2) line 35
```

Thanks.

**Reply by aregtech** [reply]
This error / warning I could fix (removed for MSVC). But there are some others and couple of questions related with this:
`WARNING: None of preprocessor defines are set. By default use LINK_AREG_API_DLL = 1. See file Switches of AREG API Core`

**Reply by aregtech** [reply]
As well the output folder is in `build`, my expectation is `product/build`

**Comment by Vikash-8090-Yadav** [comment]
Hey @aregtech and fellow dev. I am Vikash Kumar Yadav, GitHub campus Expert. It's really great to see this project and hardwork that you guys put to build it. 

I am organising one Open Source event from 18 oct to 5-10 nov. Where we have huge number of contributor who can contribute and made this project more good in terms of documentation, website  or core codebase improvement. Let me know if u r intersted, U can find my linkedin from my profile and Please give me ay medium to contact and talk more abt this. 

Thanks!!

**Comment by aregtech** [comment]
Hi @Vikash-8090-Yadav, OK. I'll contact you in LI.

**Comment by AllanOricil** [comment]
@aregtech 

In my [personal project](https://github.com/AllanOricil/esp32-mfa-totp-generator) I decided to use mqtt to enable communication between my esp and my server. The data I send through mqtt is sensitive. I'm sending MFA Secrets used to generate TOTP codes. Could you help me to justify changing it to rpc? It seems more secure but I'm not sure. 

. can I interact with the esp32 from node-red? I'm building the server that interacts with the esp32 using node-red. It would be nice to keep using it. Do you have an example to share?

. My server will be used by clients, such as chrome extension and web app, so I added authentication using keycloak in it, however I also added authentication to the mqtt topic. This way my server is the only thing that can communicate to certain topics. Is there an authentication layer in the rpc server your sdk creates?

. MQTT reconnection lock the program for up to around 5 seconds. My UI freezes when this happen. Can I overcome this issue if I switch to rpc?

**Reply by aregtech** [reply]
Hi @AllanOricil ,

I would love to help you, but I don't know much about architecture of your project and never used node-red, so that don't know what are the features of node-red and what are the limitations. I'm not sure that I'm the right person to help you.

Nevertheless, my thoughts about your question are:

1. If node-red allows bi-directional communication, i.e. it is possible in real-time mode to push messages from server to node-red, it will be easy and you need to react to events. Otherwise, I guess you need a timer on red-node that pulls messages from the server then react on the messages (if updated).
2. I don't have special authentication layer. The areg framework is a communication engine. The secured communication, such as SSL, is planned to have, but in the next versions after developing some tools for the SDK. The higher level authentications are out of areg scope, because it is user/project specific and very sensitive topic. But it is possible to integrate on top of framework as a separate service.
3. I'm not sure that this is a problem of protocol. Normally, such things happen when UI and (re-)connection run in the same thread. I would split them in a separate tasks to run in parallel (multitasking).

Best regards and good luck.
